### PR TITLE
feat: support gradient metric finalization for internal medicine monitors

### PR DESCRIPTION
**Pull request by yuweivvv**
### PR types

New features

### PR changes

Others

### Description

This PR adds optimizer-step integration for Internal Medicine gradient monitors.

Before the optimizer updates model parameters, `TrainerCallback` now calls `finalize_scaled_grad_metrics()` on registered monitors when the method is available. The current AMP `GradScaler` is passed to the monitor so that gradient metrics collected from autograd hooks, such as `h_res_logits_grad_min` and `h_res_logits_grad_max`, can be converted back to their unscaled values before being flushed.

The callback uses capability detection to remain backward-compatible with monitors that do not implement `finalize_scaled_grad_metrics()`.

This change does not modify model gradients or optimizer behavior. It only affects the values reported by Internal Medicine monitoring.

### CODE REVIEW

File: paddleformers/trainer/trainer_callback.py [hunk 1]
```diff
@@ -1045,6 +1045,15 @@ def on_step_begin(self, args, state, control, **kwargs):
...
                 collect()
 
+    def on_optimizer_begin(self, args, state, control, scaler=None, **kwargs):
+        if not self._setup_done:
+            return
+
+        for monitor in self._monitor_dict.values():
+            finalize = getattr(monitor, "finalize_scaled_grad_metrics", None)
+            if finalize is not None:
+                finalize(scaler)
```
Comment (risemeup1111) [inline paddleformers/trainer/trainer_callback.py:1055]: P1 当前锁定的 `internal_medicine` 依赖中没有这个 finalize 实现，调用不会生效。

`third_party/llm-internal-medicine` 的 gitlink 与 base 相同，所锁定版本的 `src/internal_medicine/backends/paddlefleet/__init__.py` 只注册 `qk_stats`、`moe_health`、`massive_act`，没有 `finalize_scaled_grad_metrics`；该依赖源码中也不存在这个方法。因此 `_monitor_dict` 中的 monitor 都会跳过这里，PR 标题中的 h_res_logits 梯度指标不会被 finalize 或上报。请同步包含该方法的 submodule 提交（或将实现纳入本 PR），并增加启用 monitor 的回归测试，确认这个调用实际命中。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
| 序号 | 位置 | 优先级 | 状态 |
| --- | --- | --- | --- |
| 1 | [trainer_callback.py:1055](https://github.com/PaddlePaddle/PaddleFormers/pull/4870#discussion_r3758101933) | P1 | 🚧 |

`diff --check` 与 `python -m compileall -q paddleformers/trainer/trainer_callback.py` 已通过。针对性 pytest 未能收集：环境缺少 `colorlog` 依赖。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
本次 force-push 与上一版代码 tree 完全一致，[现有 P1 线程](https://github.com/PaddlePaddle/PaddleFormers/pull/4870#discussion_r3758101933)仍未解决：锁定的 submodule 仍未包含 `finalize_scaled_grad_metrics` 实现。请按该线程同步依赖实现并补充回归测试。

**Review by zjjlivein — APPROVED** [review]
